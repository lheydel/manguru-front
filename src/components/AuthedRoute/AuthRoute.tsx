import React from 'react';
import { connect } from 'react-redux';
import { Redirect, Route, RouteProps } from 'react-router-dom';
import { ApplicationState } from '../../services/common/app.states';
import { RouteFront } from '../../utils/properties';

export interface RawAuthRouteProps extends RouteProps {
    /** The component to display */
    component: React.ComponentType<any>;

    /**
     * Specify if the user should be or should not be logged
     * (if either 'should' or 'should not' is okay, user Route from react-router instead)
     */
    shouldBe: boolean;
}

export interface AuthRouteProps extends RawAuthRouteProps {
  logged: boolean;
}

/**
 * Ensure that the user is (or is not) authenticated before accessing a given component
 */
export const RawAuthRoute: React.FC<AuthRouteProps> = ({
    component: Child,
    logged,
    shouldBe,
    ...rest
}) => {
    const redirection = shouldBe ? RouteFront.LOGIN : RouteFront.HOME;
    return (
        <Route {...rest} render={routeProps => (
            (logged && shouldBe) || (!logged && !shouldBe)
            ? <Child {...routeProps} />
            : <Redirect to={{ pathname: redirection, state: { from: routeProps.location } }} />
        )} />
    );
};

const mapStateToProps = (state: ApplicationState) => ({
    logged: state.user.login.logged,
});

export const AuthRoute = connect(mapStateToProps)(RawAuthRoute);
