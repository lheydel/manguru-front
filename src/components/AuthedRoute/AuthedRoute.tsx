import React from 'react';
import { Route, Redirect, RouteProps } from 'react-router-dom';
import { ApplicationState } from '../../services/common/app.state';
import { connect } from 'react-redux';
import { UserLoginState } from '../../services/user/reducers/user.states';
import { Cookies, withCookies } from 'react-cookie';
import { Cookie } from '../../utils/properties';
import userActionners from '../../services/user/actions/user.actionners';

export interface AuthedRouteProps extends RouteProps {
  component: any;
  cookies: Cookies;
  loginJwt: typeof userActionners.loginJwtRequest;
  loginState: UserLoginState;
}

export const RawAuthedRoute: React.FC<AuthedRouteProps> = ({
    component: Child,
    loginState: { logged, loading, error },
    loginState,
    cookies,
    loginJwt,
    ...rest
}) => {

    // if user not logged but has an Authorization cookie, try to authenticate him
    if (!logged && !error && cookies.get(Cookie.AUTH)) {
        if (!loading) {
            loginJwt();
        }
        return null;
    } // else

    console.log(loginState);

    return (
        <Route {...rest} render={props => (
            logged
            ? <Child {...props} />
            : <Redirect to={{ pathname: '/login', state: { from: props.location } }} />
        )} />
    );
};

const mapStateToProps = (state: ApplicationState) => {
  return {
    loginState: state.user.login
  };
};

const mapDispatchToProps = (dispatch: Function) => {
  return {
    loginJwt: () => dispatch(userActionners.loginJwtRequest()),
  };
};

export const AuthedRoute = withCookies(connect(mapStateToProps, mapDispatchToProps)(RawAuthedRoute));
