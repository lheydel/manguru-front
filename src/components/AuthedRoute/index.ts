import { AuthedRoute } from './AuthedRoute';
export default AuthedRoute;
export * from './AuthedRoute';
