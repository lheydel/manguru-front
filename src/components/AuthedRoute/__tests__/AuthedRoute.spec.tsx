import React from 'react';
import AuthedRoute, { AuthedRouteProps, RawAuthedRoute } from '..';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Route, Router, Redirect } from 'react-router';
import { createBrowserHistory } from 'history';
import { Provider } from 'react-redux';
import createMockStore from 'redux-mock-store';
import { initialAppState } from '../../../services/common/app.state';

describe('AuthedRoute component', () => {
    const routeProps: AuthedRouteProps = {
        path: 'blblbl',
        component: RawAuthedRoute,
        logged: true
    };

    const routerWrapper = (props: AuthedRouteProps) => {
        return mount(
            <Router history={createBrowserHistory()}>
                <RawAuthedRoute {...props}></RawAuthedRoute>
            </Router>
        );
    };

    it('should render itself and children', () => {
        const wrapper = shallow(<RawAuthedRoute {...routeProps}></RawAuthedRoute>);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render itself and children when connected', () => {
        const wrapper = mount(
            <Provider store={createMockStore()(initialAppState)}>
                <Router history={createBrowserHistory()}>
                    <AuthedRoute {...routeProps}></AuthedRoute>
                </Router>
            </Provider>
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render the component given in props when user is logged', () => {
        routeProps.logged = true;
        const wrapper = routerWrapper(routeProps);
        expect(wrapper.find(Route).props().render()).toEqual(<routeProps.component />);
    });

    it('should redirect when user is not logged', () => {
        routeProps.logged = false;
        const wrapper = routerWrapper(routeProps);
        expect(wrapper.find(Route).props().render({location: ''})).toEqual(<Redirect to={expect.anything()} />);
    });
});
