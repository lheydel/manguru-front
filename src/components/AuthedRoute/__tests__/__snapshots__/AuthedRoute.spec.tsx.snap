// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthedRoute component should render itself and children 1`] = `
<Route
  path="blblbl"
  render={[Function]}
/>
`;

exports[`AuthedRoute component should render itself and children when connected 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Connect(RawAuthedRoute)
      component={[Function]}
      logged={true}
      path="blblbl"
    >
      <RawAuthedRoute
        component={[Function]}
        dispatch={[Function]}
        logged={false}
        path="blblbl"
      >
        <Route
          dispatch={[Function]}
          path="blblbl"
          render={[Function]}
        />
      </RawAuthedRoute>
    </Connect(RawAuthedRoute)>
  </Router>
</Provider>
`;
